exonum-java-binding/core/rust/src/proxy/runtime.rs

```diff
@@ -0,0 +1,596 @@
+/*
+ * Copyright 2019 The Exonum Team
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+use std::collections::HashMap;
+use std::collections::HashSet;
+use std::fmt;
+use std::str::FromStr;
+use exonum::{
+    api::ApiContext,
+    blockchain::Schema as CoreSchema,
+    crypto::{Hash, PublicKey, SecretKey},
+    helpers::{Height, ValidatorId},
+    node::ApiSender,
+    proto::Any,
+    runtime::{
+        api::ServiceApiBuilder,
+        ArtifactId, ArtifactProtobufSpec, CallInfo,
+        dispatcher::{Dispatcher, DispatcherSender},
+        ErrorKind, ExecutionContext, ExecutionError, InstanceDescriptor, InstanceId, InstanceSpec,
+        Runtime, RuntimeIdentifier, StateHashAggregator
+    },
+    messages::BinaryValue,
+};
+use exonum_merkledb::{Fork, Snapshot};
+use futures::{Future, IntoFuture};
+use jni::{
+    Executor,
+    objects::{GlobalRef, JValue, JObject},
+    signature::JavaType,
+};
+use JniErrorKind;
+use JniResult;
+use Handle;
+use proxy::node::NodeContext;
+use proto;
+use semver::Version;
+use storage::View;
+use to_handle;
+use utils::{
+    jni_cache::runtime_adapter,
+    panic_on_exception, unwrap_jni,
+};
+
+/// A proxy for `ServiceRuntimeAdapter`s.
+#[derive(Clone)]
+pub struct JavaRuntimeProxy {
+    exec: Executor,
+    runtime_adapter: GlobalRef,
+    deployed_artifacts: HashSet<JavaArtifactId>,
+    started_services: HashMap<InstanceId, Instance>,
+    started_services_by_name: HashMap<String, InstanceId>,
+}
+
+/// Service identification properties within `JavaRuntimeProxy`
+#[derive(Debug, Clone)]
+struct Instance {
```

Comment:
I believe something similar (InstanceId?) exists in Core, doesn't it?

Reply:
There is similar structure in core - ```InstanceSpec```, which is extended by ```ArtifactId```. 
```InstanceSpec``` or simple ```InstanceId```  (integer identifier) can be used instead of custom struct ```Instance ```, the second variant is available if mapping services by name is not required for future.